### ISSUE TITLE
`BitVec::append` fails silently

### ISSUE THREAD

**Issue by XAMPPRocky** [contain-rs/bit-vec#63]
Hello, I've been using BitVec extensively for a project and I noticed that `BitVec::append` doesn't work all the time. See below for a minimal example, where I would expect the output to be `[01, 00, 02, 03]`, but `BitVec` returns `[01, 00, 02, 00]`.

```rust
use bit_vec::BitVec;

#[test]
fn it_works() {
    // Create a 8 bit integer with value (`1`) encoded in BE.
    let mut a = BitVec::from_elem(8, false);
    a.set(7, true);

    // Create a 16 bit integer with value (`2`) encoded in BE.
    let mut b = BitVec::from_elem(16, false);
    b.set(14, true);

    // Create a 8 bit integer with value (`3`) encoded in BE.
    let mut c = BitVec::from_elem(8, false);
    c.set(6, true);
    c.set(7, true);

    // Merge all bit_vecs together
    a.append(&mut b);
    a.append(&mut c);

    assert_eq!(&[01, 00, 02, 03][..], &*a.to_bytes());
}
```

**Comment by Gankra** [comment]
yeah i'm pretty sure this code is just wrong: https://github.com/contain-rs/bit-vec/blob/master/src/lib.rs#L962-L971

I can't remember the exact layout/constraints of bitvec, but these two things contradict each other:

(terminology: an "out of bounds" block is one which is pushed in the Vec's representation, but which contains no bits in the BitVec's representation (so vec_index * B > bit_vec_len))

1. The loop only ever mutating `self.storage.last_mut()` assumes that you never have out of bounds blocks (grabbing the last block must 

2. The unconditional `self.storage.push(block >> (B::bits() - b));`  in the loop is assuming that `other.len()` is large enough that we always need to push another block. When this is not the case, we push an out of bounds block.

So your first append "corrupts" the bitvec for the second append. If you convert your final BitVec to a Vec, you should find your missing bits in the second "block".

The behaviour of 2 is "bad" regardless since it does unnecessary work and may unnecessarily allocate. We should have a check for how big `self.len() % B + other.len() % B` is, and on the last iteration not do a push. 

Whether this code needs to also handle out of bounds blocks is less clear to me (but can easily be fixed by just getting the last *in bounds* block, and as an optimization, not pushing unless you're out of out-of-bounds blocks).

**Comment by woodruffw** [comment]
I also observed this behavior in some code that called `append` multiple times, causing silent corruption of the `BitVec`. Any chance of a fix landing?

**Comment by j-schwar** [comment]
I also came across this today. Since the PR for this is over half a year old and didn't seem to be going anywhere, I tried my hand at fixing it. Would you like me to submit a PR with my changes to try and get this fixed?

**Comment by pczarn** [comment]
Fixed in 0.6.3 with a PR from @j-schwar . in #68
